File: algorithms/binsearch/binsearch.cpp
#include <iostream>

using namespace std;

template <class T>
T binarySearch(T *arr, int l, int r, T key)
{
    int mid = (l+r)/2;
    cout<<l<<" "<<mid<<" "<<r<<"\n";
    cout<<arr[mid]<<" "<<key<<"\n";

    if(arr[mid]==key)
        return mid;
    
    if(l>=r)
        return -1;

    if(arr[mid]>key)
        return binarySearch(arr, l , mid-1, key);
    else
        return binarySearch(arr, mid+1, r, key);
    
}

int main()
{
    int arr[] = {2, 3, 5, 6, 8, 9, 14};
    int n = sizeof(arr) / sizeof(arr[0]);

    cout << binarySearch(arr, 0, n - 1, 1) << "\n";
}

File: algorithms/binsearch/binsearch.py
def binarySearch(arr, l, r, key):
    mid = int((l+r)/2)
    print(l, mid, r)

    if(arr[mid] == key):
        print(arr[mid], key)
        return mid
    
    if(l==r):
        return -1

    if(arr[mid] > key):
        return binarySearch(arr, l, mid-1, key)
    else:
        return binarySearch(arr, mid+1, r, key)

def main():
    arr = [2, 3, 5, 6, 8, 9, 14]
    print(arr)
    print(binarySearch(arr, 0, len(arr)-1, 5))


if __name__ == "__main__":
    main()
